### ISSUE TITLE
DX: Code coverage feels broken due to only re-running one test file

### ISSUE THREAD

**Issue by lydell** [vitest-dev/vitest#4347]
### Describe the bug

When requiring 100 % code coverage (for example), vitest feels broken: It sometimes says that the coverage goals weren’t met, but they are! This happens when you edit a test while `vitest` is running and watching, which seems to trigger only that test to run, which results in – technically – not all source code being run (unless every test file tests everything). This is confusing to me. The tool is reporting a problem that isn’t a problem. This makes vitest feel broken to me: I can’t really trust what I see after I save a file – I need to manually trigger a run for it to “work”.

A potential solution is to print something like “Not all tests were run, so coverage was not considered. Press Enter to run all tests and check coverage.”

### Reproduction

1. Clone https://github.com/lydell/vitest-quirk

2. `npm cit`

   ```
   added 92 packages, and audited 93 packages in 597ms

   18 packages are looking for funding
     run `npm fund` for details

   found 0 vulnerabilities

   > test
   > vitest


    DEV  v0.34.6 /Users/simon/stuff/vitest-quirk
         Coverage enabled with v8

    ✓ test1.test.ts (1)
    ✓ test2.test.ts (1)

    Test Files  2 passed (2)
         Tests  2 passed (2)
      Start at  22:43:18
      Duration  133ms (transform 24ms, setup 0ms, collect 20ms, tests 4ms, environment 0ms, prepare 88ms)

    % Coverage report from v8
   ----------|---------|----------|---------|---------|-------------------
   File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
   ----------|---------|----------|---------|---------|-------------------
   All files |     100 |      100 |     100 |     100 |
    index.ts |     100 |      100 |     100 |     100 |
   ----------|---------|----------|---------|---------|-------------------

    PASS  Waiting for file changes...
          press h to show help, press q to quit
   ```

   ✅ 100 % code coverage.

3. In another terminal, run `touch test1.test.ts`

   ```
    RERUN  test1.test.ts x1

    ✓ test1.test.ts (1)
      ✓ myFunction1

    Test Files  1 passed (1)
         Tests  1 passed (1)
      Start at  22:44:57
      Duration  9ms

    % Coverage report from v8
   ----------|---------|----------|---------|---------|-------------------
   File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
   ----------|---------|----------|---------|---------|-------------------
   All files |     100 |      100 |      50 |     100 |
    index.ts |     100 |      100 |      50 |     100 |
   ----------|---------|----------|---------|---------|-------------------
   ERROR: Coverage for functions (50%) does not meet global threshold (100%)

    PASS  Waiting for file changes...
          press h to show help, press q to quit
   ```

   🤔 Suddenly I only have 50 % coverage?

4. Workaround: Press `Enter` in the vitest terminal.

   ```
    RERUN

    ✓ test2.test.ts (1)
    ✓ test1.test.ts (1)

    Test Files  2 passed (2)
         Tests  2 passed (2)
      Start at  22:45:58
      Duration  14ms

    % Coverage report from v8
   ----------|---------|----------|---------|---------|-------------------
   File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
   ----------|---------|----------|---------|---------|-------------------
   All files |     100 |      100 |     100 |     100 |
    index.ts |     100 |      100 |     100 |     100 |
   ----------|---------|----------|---------|---------|-------------------

    PASS  Waiting for file changes...
          press h to show help, press q to quit
   ```

   ✅ Now it reports 100 % coverage again.

This makes vitest feel broken to me: I can’t really trust what I see after I save a file – I need to manually trigger a run for it to “work”.

### System Info

```shell
System:
    OS: macOS 13.6
    CPU: (10) arm64 Apple M1 Pro
    Memory: 186.52 MB / 32.00 GB
    Shell: 3.6.1 - /opt/homebrew/bin/fish
  Binaries:
    Node: 20.5.0 - ~/.local/share/nvm/v20.5.0/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 9.8.0 - ~/.local/share/nvm/v20.5.0/bin/npm
    pnpm: 8.8.0 - /opt/homebrew/bin/pnpm
  Browsers:
    Chrome: 118.0.5993.96
    Safari: 16.6
  npmPackages:
    @vitest/coverage-v8: 0.34.6 => 0.34.6
    vitest: 0.34.6 => 0.34.6
```
```


### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by AriPerkkio** [comment]
> 2. ✅ 100 % code coverage.
> 3. In another terminal, run touch `test1.test.ts`
> 🤔 Suddenly I only have 50 % coverage?

Works as expected. 

Coverage report includes coverage generated by the test files that are run. If you run only a subset of your test files, you cannot see coverage of all test files. As far as I know, every (Javascript) code coverage tool works like this. You cannot see coverage of test files that are not run.

**Comment by lydell** [comment]
> Works as expected.

Yeah, it’s expected when you think about it. But until then it feels broken. Stupid humans running computers 😄 

> As far as I know, every (Javascript) code coverage tool works like this.

Then let’s be better than all the other tools! Feels on-brand for vitest 😄

**Comment by AriPerkkio** [comment]
Watch mode re-runs only the test files that are changed or that are affected by the changed source files. In order to show coverage report of all test files, all test files need to be run. Running all tests when a single file changes during watch mode makes no sense as it would make developer experience worse.

Any suggestions what to change here?

**Comment by lydell** [comment]
Quoting from first post:

> A potential solution is to print something like “Not all tests were run, so coverage goals were not considered. Press Enter to run all tests and check coverage.”

My first idea was to not show coverage _at all_ when not all tests are run, but now I realize that people might use the output to quickly iterate on the coverage of a certain file via a certain test. So maybe it makes sense to print the table, but not the `ERROR: Coverage for functions (50%) does not meet global threshold (100%)` lines. It could look like this (see the added NOTE line):

```
 RERUN  test1.test.ts x1

 ✓ test1.test.ts (1)
   ✓ myFunction1

 Test Files  1 passed (1)
      Tests  1 passed (1)
   Start at  22:44:57
   Duration  9ms

 % Coverage report from v8
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
----------|---------|----------|---------|---------|-------------------
All files |     100 |      100 |      50 |     100 |
 index.ts |     100 |      100 |      50 |     100 |
----------|---------|----------|---------|---------|-------------------
NOTE: Not all tests were run, so global threshold was not checked. Press Enter to run all tests and check coverage thresholds.

 PASS  Waiting for file changes...
       press h to show help, press q to quit
```

Feel free to change the wording of the `NOTE` line – the point here is that something along those lines is more helpful to the user than `ERROR` that you need to learn to ignore and not ignore depending on the situation.

**Comment by AriPerkkio** [comment]
> My first idea was to not show coverage at all when not all tests are run, but now I realize that **<ins>people might use the output to quickly iterate on the coverage of a certain file</ins>** via a certain test. So maybe it makes sense to print the table, but not the `ERROR: Coverage for functions (50%) does not meet global threshold (100%)` lines.

They may also be using the coverage threshold information when iterating on the tests - especially if they are using `coverage.perFile` option. It still makes sense when a single test file is used to cover source files enough.

I think if we did not show coverage threshold information during watch mode re-runs we would get bug reports stating we should. Especially since other tools (like Jest) do it too.

If you want to disable coverage thresholds like this on your project by patching Vitest, you can change if-block's condition below. There's already `allTestsRun` variable in that scope which indicates wheter all teste were run.

https://github.com/vitest-dev/vitest/blob/4293e1b15111316b8ef622d04f4e5d8b70ad9474/packages/coverage-v8/src/provider.ts#L157-L171

Or simply not use thresholds in watch mode at all:
```json
// package.json
"test": "vitest run --coverage.enabled --coverage.100",
"test:watch": "vitest watch",
```

**Comment by lydell** [comment]
Here’s how I ended up creating this issue.

I was migrating from Jest. In Jest, I run `jest --watch` while developing and have a separate script that runs `jest --coverage` like you suggested. I tried replicating that in vitest, but didn’t manage to get the coverage flags working (#4307) so I went with vitest.config.ts instead.

I ran `vitest` in my project, and all was well. Tests passed, coverage was 100 %. I then made a small tweak to an error message in the library I was working on. vitest then showed me that some snapshots were out-of-date. They looked good, so I pressed `u` to update them.

Then vitest told me I didn’t have 100 % coverage anymore, and I was confused. It didn’t make sense – the change I had made shouldn’t have resulted in less coverage … or should it? But then I did the classic solve-all-solution: I pressed Enter to run the tests again. Then they passed.

This then happened a couple of times more, and I still couldn’t figure out why. Knowing that vitest is 0.x, I was thinking maybe there’s a bug around watch mode and code coverage (I’ve written watch modes for CLI tools myself, and they can be pretty tricky).

After a couple of days I finally understood what was happening. When pressing `u` to update a snapshot, vitest writes to some `.test.ts` files and then its watcher picks up those changes and only re-runs those files, and then all source code isn’t run.

In Jest, I have always used `jest --coverage` because that’s what I read somewhere once. I haven’t even checked if it’s possible to turn on coverage always in the config file. Now that I had tested having it on always in vitest, I kinda liked it. It’s easy to miss some coverage otherwise during development.

I know that vitest cares a lot about DX – it is the most delightful test runner I’ve ever used – so I thought I’d open this issue to let you know. So now you have the full story from a beginner, and it’s up to you if you think this is good enough, or if you’d like to go the extra mile.

**Comment by AriPerkkio** [comment]
I think we could display an informative message above coverage threshold messages to indicate that not all tests were run so thresholds may be incorrect. We should not disable coverage thresholds completely in these cases as they provide very good feedback to developers who have only a single test covering the file. 

```diff
 % Coverage report from v8
|----------|---------|----------|---------|---------|-------------------
 File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
|----------|---------|----------|---------|---------|-------------------
 All files |     100 |      100 |      50 |     100 |
  index.ts |     100 |      100 |      50 |     100 |
|----------|---------|----------|---------|---------|-------------------

+NOTE: Only the changed tests were run. Coverage thresholds do not include coverage of all test files.
ERROR: Coverage for functions (50%) does not meet global threshold (100%)

 PASS  Waiting for file changes...
       press h to show help, press q to quit
```

Or maybe even something like `${testCount}/${allTests}` if we have that information available easily:

```
NOTE: Only 1/52 tests were run. Coverage thresholds do not include coverage of all test files.
```

**Comment by lydell** [comment]
That looks good! :+1:

**Comment by robintown** [comment]
> In order to show coverage report of all test files, all test files need to be run.

Completely naively, if my test suites run in isolation and each of them covers a certain set of statements, the full coverage report is given by the union of all of these sets. So, it seems possible in theory that, when I edit a test suite, my test runner could re-run just that one test and take the union of [its new coverage report] + [the cached results of the other tests from the previous run] to produce a full, up-to-date coverage report.

Similarly, if my test runner can tell which tests depend on which modules, as Vitest does, then when I edit a module, it should only need to recalculate the partial coverage reports of the tests that have been re-run.

I don't know what the memory overhead of remembering which tests ran which lines would look like, and I also don't know whether coverage reporters expose that information at all :sweat_smile: But that would be the ideal developer experience, I think!

**Comment by AriPerkkio** [comment]
Yep, sounds possible in theory. At the moment we don't know test file specific coverage so preserving previous results will cause incorrect results when they are merged. 

This is where the results are collected:

https://github.com/vitest-dev/vitest/blob/c321a3fd29902fa9295fe219816381884c40a8cd/packages/vitest/src/runtime/runners/index.ts#L86-L99

To preserve the results from previous run you can use [`coverage.cleanOnRerun`](https://vitest.dev/config/#coverage-cleanonrerun) in watch-mode.

https://github.com/vitest-dev/vitest/blob/c321a3fd29902fa9295fe219816381884c40a8cd/packages/coverage-v8/src/provider.ts#L278-L279

**Comment by AriPerkkio** [comment]
This was fixed by https://github.com/vitest-dev/vitest/pull/6592.

You can set [`coverage.cleanOnReRun`](https://vitest.dev/config/#coverage-cleanonrerun) to `false`. Vitest will then merge the previous run's coverage results with the last one. You can keep changing a single file while seeing all other files covered.

**Comment by lydell** [comment]
Awesome, well done!
